src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs

```diff
@@ -0,0 +1,1553 @@
+        internal static byte[] BoundaryPrefix = new byte[] { (byte)'\r', (byte)'\n', (byte)'-', (byte)'-' };
+    }
+
+    internal enum MimeWriterState
+    {
+        Start,
+        StartPreface,
+        StartPart,
+        Header,
+        Content,
+        Closed,
+    }
+
+    internal class MimeWriter
+    {
+        private Stream stream;
+        private byte[] boundaryBytes;
+        private MimeWriterState state;
+        private BufferedWrite bufferedWrite;
+        private Stream contentStream;
+
+        internal MimeWriter(Stream stream, string boundary)
+        {
+            if (stream == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(stream));
+            if (boundary == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(boundary));
+
+            this.stream = stream;
+            boundaryBytes = MimeWriter.GetBoundaryBytes(boundary);
+            state = MimeWriterState.Start;
+            bufferedWrite = new BufferedWrite();
+        }
+
+        internal static int GetHeaderSize(string name, string value, int maxSizeInBytes)
+        {
+            if (name == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(name));
+            if (value == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(value));
+
+            int size = XmlMtomWriter.ValidateSizeOfMessage(maxSizeInBytes, 0, MimeGlobals.COLONSPACE.Length + MimeGlobals.CRLF.Length);
+            size += XmlMtomWriter.ValidateSizeOfMessage(maxSizeInBytes, size, name.Length);
+            size += XmlMtomWriter.ValidateSizeOfMessage(maxSizeInBytes, size, value.Length);
+            return size;
+        }
+
+        internal static byte[] GetBoundaryBytes(string boundary)
+        {
+            byte[] boundaryBytes = new byte[boundary.Length + MimeGlobals.BoundaryPrefix.Length];
+            for (int i = 0; i < MimeGlobals.BoundaryPrefix.Length; i++)
+                boundaryBytes[i] = MimeGlobals.BoundaryPrefix[i];
+            Encoding.ASCII.GetBytes(boundary, 0, boundary.Length, boundaryBytes, MimeGlobals.BoundaryPrefix.Length);
+            return boundaryBytes;
+        }
+
+        internal MimeWriterState WriteState
+        {
+            get
+            {
+                return state;
+            }
+        }
+
+        internal int GetBoundarySize()
+        {
+            return boundaryBytes.Length;
+        }
+
+        internal void StartPreface()
+        {
+            if (state != MimeWriterState.Start)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.MimeWriterInvalidStateForStartPreface, state.ToString())));
+
+            state = MimeWriterState.StartPreface;
+        }
+
+        internal async Task StartPartAsync()
```

Comment:
Nit: Unlike all the other added async methods, this one along with FlushAsync and GetContentStreamAsync appear _before_ their synchronous counterparts.

Reply:
Fixing, that is annoying 😄